build: Require pkg-config for all of the hosts

hebasto opened the pull request:
This PR:
- is based on #18297 (already merged)
- drops all of the non-pkg-config paths from the `configure` script

Ref: #17768

hebasto committed:
build: Always define ZMQ_STATIC for MinGW

hebasto committed:
build: Always use pkg-config

hebasto committed:
build: Drop dead non-pkg-config code for UNIVALUE check

hebasto committed:
build: Fix indentation in UNIVALUE check

hebasto committed:
build: Drop dead non-pkg-config code for ZMQ check

hebasto committed:
build: Drop dead non-pkg-config code for qrencode check

hebasto committed:
build: Drop dead non-pkg-config code for libevent check

laanwj commented:
I like this idea, especially the simplification of the autoconf scripts. Concept ACK.

hebasto commented:
Updated 284418ad2446fec21546e485ec0af0edde2eb773 -> b3b769adf1868b2c3ebdd1d9778fac58125e2b8d ([pr18307.01](https://github.com/hebasto/bitcoin/commits/pr18307.01) -> [pr18307.02](https://github.com/hebasto/bitcoin/commits/pr18307.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18307.01..pr18307.02)):

- rebased due to the conflict with #18298

theuni commented:
Concept ACK, thanks for working on this. Will review the chain.

hebasto commented:
Rebased b3b769adf1868b2c3ebdd1d9778fac58125e2b8d -> a4535bfd4d47538db47b9787da8922cd06bd6fd7 ([pr18307.02](https://github.com/hebasto/bitcoin/commits/pr18307.02) -> [pr18307.03](https://github.com/hebasto/bitcoin/commits/pr18307.03)) due to the merging of #18297.

hebasto commented:
@fanquake 
May I ask "Needs gitian build" and "Needs Guix build" labels for this PR?

fanquake reviewed (commented):
Concept ACK - This looks pretty good, and is a nice simplification. This should be a no-op for most platforms, besides some reordering of package checks in configure. Obviously Windows is different, and after this PR, should be full pkg-config. This is the main difference I'm seeing in `config.log`:
```diff
# libevent
-EVENT_CFLAGS=''
-EVENT_LIBS='-levent'
+EVENT_CFLAGS='-I/bitcoin/depends/x86_64-w64-mingw32/include'
+EVENT_LIBS='-L/bitcoin/depends/x86_64-w64-mingw32/lib -levent -lws2_32'
# qrencode
-QR_CFLAGS=''
-QR_LIBS='-lqrencode'
+QR_CFLAGS='-I/bitcoin/depends/x86_64-w64-mingw32/include'
+QR_LIBS='-L/bitcoin/depends/x86_64-w64-mingw32/lib -lqrencode -lpthread'
# zeromq
-ZMQ_CFLAGS=' -DZMQ_STATIC'
-ZMQ_LIBS='-lzmq'
+ZMQ_CFLAGS='-I/bitcoin/depends/x86_64-w64-mingw32/include -DZMQ_STATIC'
+ZMQ_LIBS='-L/bitcoin/depends/x86_64-w64-mingw32/lib -lzmq -liphlpapi -lpthread'
```

`liphlpapi` was added to libzmqs `PKGCFG_LIBS_PRIVATE`, when static linking, in https://github.com/zeromq/libzmq/pull/2787.

hebasto commented on configure.ac:
```diff
@@ -1301,115 +1292,66 @@ CPPFLAGS="$TEMP_CPPFLAGS"
+if test x$build_bitcoin_cli$build_bitcoind$bitcoin_enable_qt$use_tests$use_bench != xnonononono; then
+  PKG_CHECK_MODULES([EVENT], [libevent >= 2.0.21], [use_libevent=yes], [AC_MSG_ERROR(libevent version 2.0.21 or greater not found.)])
+  if test x$TARGET_OS != xwindows; then
+    PKG_CHECK_MODULES([EVENT_PTHREADS], [libevent_pthreads],, AC_MSG_ERROR([libevent_pthreads not found]))
```
> Was removing the version checking for `libevent_pthreads` intentional?

No, it seems like a rebasing glitch. Going to fix.
Thanks!

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18307#issuecomment-647371783).

hebasto commented:
Updated a4535bfd4d47538db47b9787da8922cd06bd6fd7 -> 4e006b640eb94d0699169b114842aca3b6ce585a ([pr18307.03](https://github.com/hebasto/bitcoin/commits/pr18307.03) -> [pr18307.05](https://github.com/hebasto/bitcoin/commits/pr18307.05), [diff](https://github.com/hebasto/bitcoin/compare/pr18307.03..pr18307.05)):

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/18307#discussion_r443360536):
> Was removing the version checking for `libevent_pthreads` intentional?

hebasto committed:
build: Drop use_pkgconfig check for libmultiprocess check

hebasto committed:
build: Drop unused use_pkgconfig variable

hebasto committed:
build: Detect missed pkg-config early

hebasto commented:
Updated 4e006b640eb94d0699169b114842aca3b6ce585a -> 92bc268e4af4ebcbde08567ea00e019ac509a769 ([pr18307.05](https://github.com/hebasto/bitcoin/commits/pr18307.05) -> [pr18307.06](https://github.com/hebasto/bitcoin/commits/pr18307.06), [diff](https://github.com/hebasto/bitcoin/compare/pr18307.05..pr18307.06)):

- fixed conflict with #18677

laanwj commented:
ACK 92bc268e4af4ebcbde08567ea00e019ac509a769
Changes look correct, and I checked that there are no remaining matches for `use_pkgconfig` after this.

laanwj commented on configure.ac:
```diff
@@ -1301,115 +1292,66 @@ CPPFLAGS="$TEMP_CPPFLAGS"
+if test "x$use_zmq" = xyes; then
+  dnl Assume libzmq was built for static linking
+  case $host in
+    *mingw*)
```
It's somewhat unfortunate that this mingw-specific workaround is still ncessary, that the required flag is not passed in through the pkgconfig file in the case of a static build. But I suppose this is an upstream issue.

hebasto replied:
Related to #8238.

fanquake reviewed (approved):
ACK 92bc268e4af4ebcbde08567ea00e019ac509a769. I re-gitian-built. There are a couple follow-ups that I'll PR shortly. Thanks for addressing my feedback above. I took too long to get back to this.
